Added different error messages to table writer

janbridley opened the pull request:
## Description

- Improved error logic for setting ``LoggerCategories`` in ``hoomd.write.Table``.
- Added new error message that details if invalid categories have been set


## Motivation and context

Prior to this change, users adding disallowed categories to ``hoomd.write.Table`` would receive an error intended for when ``LoggerCategories.scalar`` was not set. This fix updates the error logic to raise a more descriptive error detailing which disallowed categories have been set by the user.

Resolves #1510 

## How has this been tested?

Changes were tested with every permutation of LoggerCategories, which return the expected errors.

## Change log

```
Fixed:

* hoomd.write.Table now raises correct error if invalid LoggerCategories are set.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

janbridley committed:
Added different error messages to table writer

Prior to this, adding disallowed LoggerCategories for TableWriter returned an error "Given Logger must have the scalar categories set". The error has been updated to correctly identify whether the scalar category is set, and lists disallowed categories that the user has set.

janbridley committed:
Updated Credits.rst

janbridley committed:
Reformatted to comply with required line length

tommy-waltmann commented on hoomd/write/table.py:
```diff
@@ -208,12 +208,16 @@ def __init__(self,
-                or logger.categories & self._invalid_logger_categories
-                !=  # noqa: W504 (yapf formats this incorrectly
-                LoggerCategories.NONE):
+        if LoggerCategories.scalar not in logger.categories:
+            raise ValueError("Given Logger must have the scalar categories set.")
+        elif (
+            logger.categories & self._invalid_logger_categories != LoggerCategories.NONE
+        ):
```
Is the boolean expression in the elif condition of the if-elif general enough to cover the boolean expression in the if condition?

janbridley replied:
Yes, as far as I am aware: the cases are as follows.

**If scalar category is not set:** the logger fails. This is in discussion [below](https://github.com/glotzerlab/hoomd-blue/pull/1519#discussion_r1138716821), but would work the same way if the case checked for scalar OR string

**If there are logger categories AND invalid logger categories:** the logger fails as intended.

**Other cases:**

- **There are no logger categories:** logger passes, but provides no output (expected)
- **There are logger categories but no invalid categories:** logger passes.
- **There are ONLY invalid categories:** this is the same as the first case, and will fail because there is no scalar category set.
- **There is only the string category:** this is also the same as the first case, and fails because scalar is not set

joaander replied:
It would be awesome if you added a pytest unit test that checks these.

janbridley replied:
Added - see [68df2a3](https://github.com/glotzerlab/hoomd-blue/pull/1519/commits/68df2a30053109ba1eef7edd49e10a6cfcc373a4)

joaander replied:
Thanks!

tommy-waltmann reviewed (commented):
Great work @janbridley . I wanted to bring one point up for consideration which may make the if-elif logic simpler.

joaander commented on hoomd/write/table.py:
```diff
@@ -208,12 +208,16 @@ def __init__(self,
-                or logger.categories & self._invalid_logger_categories
-                !=  # noqa: W504 (yapf formats this incorrectly
-                LoggerCategories.NONE):
+        if LoggerCategories.scalar not in logger.categories:
```
Why not accept a logger that only takes `string` categories?

janbridley replied:
I'm not sure - I retained the original behavior (that required ``scalar`` categories to be set) but I don't see why a ``string``-only logger wouldn't work.  @b-butler, do you have any insight into why it was initially set up this way? If you don't recall any issues, I will refine the logic to allow ``string``-only loggers.

b-butler replied:
We should. I just didn't originally for who knows why. @janbridley please change this here to fix my past mistake.

janbridley replied:
Done! See [d4a292b](https://github.com/glotzerlab/hoomd-blue/pull/1519/commits/d4a292b17cfa118d77cffee53f558b0ffd7ea953)

janbridley committed:
Updated write.Table invalid categories to allow string-only loggers

janbridley committed:
Reformatted table.py to be compatible with flake8 formatting

janbridley committed:
Added pytest unit test for writer.Table

janbridley committed:
Improved comments and reformatted with yapf

janbridley committed:
Fixed error with logical comparison of LoggerCategories

Applying logical operators to LoggerCategories can cause an addition that results in a new LoggerCategories.any object. This caused unexpected error behavior, which has been fixed.

janbridley committed:
Parameterized new pytest and fixed test logic

janbridley committed:
Cleaned erroneous print() call in test_table

janbridley committed:
Clarified logic and errors, corrected parameterization

Parameterization was performed on an incomplete input set, as the first LoggerCategory was not correctly split. Regex fixes this issue.

joaander reviewed (approved):
Thanks, this looks good to me!

b-butler commented on hoomd/write/table.py:
```diff
@@ -205,15 +205,28 @@ def __init__(self,
-                or logger.categories & self._invalid_logger_categories
-                !=  # noqa: W504 (yapf formats this incorrectly
-                LoggerCategories.NONE):
+        # If logger.categories == none, pass automatically
```
This comment just repeats the next line. Clarification comments like this are good when the next line(s) is/are difficult to parse or call some obscure function.
```suggestion
```

b-butler commented on hoomd/write/table.py:
```diff
@@ -205,15 +205,28 @@ def __init__(self,
+                LoggerCategories.scalar, LoggerCategories.string
+        ]) & logger.categories == LoggerCategories.NONE:
+            raise ValueError(
+                "Table Logger may only have scalar or string categories set.")
+        elif _invalid_inputs != LoggerCategories.NONE:
+            # If logger.categories != NONE passes:
+            # - Logger is empty (no invalid categories)
+            # If invalid_categories != NONE passes:
+            # - No invalid categories are set
             raise ValueError(
-                "Given Logger must have the scalar categories set.")
+                "{} are incompatible with write.Table: use write.GSD instead."
+                .format(_invalid_inputs))
```
I still don't see the need for two warnings at this point maybe something like `Table Logger may only have scalar or string categories set. Use hoomd.write.GSD for other categories.`. Also if a user only asks to log invalid categories, the error does not mention this but instead tells them to add valid categories.

janbridley replied:
If we want to condense to a single error message, we could use the following code: 
 
```python
# Generate LoggerCategories for valid and invalid categories
_valid_categories = LoggerCategories.any(
    [LoggerCategories.scalar, LoggerCategories.string])
_invalid_inputs = logger.categories & self._invalid_logger_categories

if logger.categories == LoggerCategories.NONE:
    pass
elif (_valid_categories ^ LoggerCategories.ALL
      ) & logger.categories == LoggerCategories.NONE:
    pass
else:
    raise ValueError(
        "Table Logger may only have scalar or string categories set. \
            Use hoomd.write.GSD for {}.".format(_invalid_inputs))
```

~~This may be possible without converting the input categories to string...~~

Code has been updated to directly compare logger categories!

tommy-waltmann replied:
I don't have a strong preference either way on this.

b-butler commented:
I requested changes to prevent a pre-mature merging as I wanted to get @joaander and @tommy-waltmann thoughts on my comment above.

joaander commented:
> I requested changes to prevent a pre-mature merging as I wanted to get @joaander and @tommy-waltmann thoughts on my comment above.

I don't have a preference. The docs and tutorials are very clear, but if a user finds more detailed error messages helpful at the cost of a few lines of Python code then I'm fine with that.

janbridley committed:
Removed redundant comment in write.Table

Co-authored-by: Brandon Butler <butlerbr@umich.edu>

janbridley committed:
Combined errors to single unified output

Raising multiple errors depending on input data could add complexity in interpreting errors. The logic has been streamlined to raise a single, descriptive error that describes the intended function as well as current invalid inputs.
